contracts/serverless-v2/Executors.sol

```diff
@@ -237,44 +237,75 @@ contract Executors is
-        emit ExecutorRegistered(_executor, _enclaveAddress);
+        emit ExecutorRegistered(_executor, _jobCapacity);
     }
 
-    function _deregisterExecutor(
+    function _drainExecutor(
         address _executor
     ) internal {
-        if(!executors[_executor].status)
-            revert ExecutorsAlreadyInitiatedDeregister();
+        if(executors[_executor].draining)
+            revert ExecutorsAlreadyDrained();
 
-        executors[_executor].status = false;
+        executors[_executor].draining = true;
 
         // remove node from the tree
         _deleteIfPresent(_executor);
 
-        if(executors[_executor].activeJobs == 0)
-            _completeDeregister(_executor);
-        else
-            emit ExecutorDeregisterInitiated(_executor);
+        emit ExecutorsExecutorDrained(_executor);
+    }
+
+    function _reviveExecutor(
+        address _executor
+    ) internal {
+        Executor memory executorNode = executors[_executor];
+        if(!executorNode.draining)
+            revert ExecutorsAlreadyRevived();
+
+        executors[_executor].draining = false;
+
+        // insert node in the tree
+        if(executorNode.stakeAmount >= MIN_STAKE_AMOUNT && 
+            executorNode.activeJobs != executorNode.jobCapacity
+        ) {
+            _insert_unchecked(_executor, uint64(executorNode.stakeAmount));
+        }
+
+        emit ExecutorsExecutorRevived(_executor);
+    }
+
+    function _deregisterExecutor(
+        address _executor
+    ) internal {
+        if(!executors[_executor].draining)
+            revert ExecutorsNotDrained();
+        if(executors[_executor].activeJobs != 0)
+            revert ExecutorsHasPendingJobs();
+        
+        _removeStake(_executor, executors[_executor].stakeAmount);
+
+        _revokeEnclaveKey(executors[_executor].enclaveAddress);
+        delete executors[_executor];
+
+        emit ExecutorDeregistered(_executor);
     }
 
     function _addExecutorStake(
         uint256 _amount,
         address _executor
     ) internal {
-        if(!executors[_executor].status)
-            revert ExecutorsAlreadyInitiatedDeregister();
-        if(executors[_executor].unstakeStatus)
-            revert ExecutorsAlreadyInitiatedUnstake();
-        
-        uint256 prevStake = executors[_executor].stakeAmount;
+        Executor memory executorNode = executors[_executor];
+        uint256 prevStake = executorNode.stakeAmount;
         uint256 updatedStake = prevStake + _amount;
 
-        if(updatedStake >= MIN_STAKE_AMOUNT) {
+        if(
+            !executorNode.draining && 
+            executorNode.activeJobs != executorNode.jobCapacity && 
+            updatedStake >= MIN_STAKE_AMOUNT
+        ) {
             if(prevStake < MIN_STAKE_AMOUNT)
                 _insert_unchecked(_executor, uint64(updatedStake));
-            else if(executors[_executor].activeJobs != executors[_executor].jobCapacity)
+            else if(executorNode.activeJobs != executorNode.jobCapacity)
```

Comment:
this `if` condition is redundant, because it's already been checked at L303. You can just use `else`.